lib/Transforms/Secretize/WrapGeneric.cpp

```diff
@@ -0,0 +1,93 @@
+#include <iostream>
+
+#include "include/Dialect/Secret/IR/SecretDialect.h"
+#include "include/Dialect/Secret/IR/SecretOps.h"
+#include "include/Dialect/Secret/IR/SecretTypes.h"
+#include "include/Transforms/Secretize/Passes.h"
+#include "mlir/include/mlir/Dialect/Func/IR/FuncOps.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/IRMapping.h"             // from @llvm-project
+#include "mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/Passes.h"  // from @llvm-project
+
+namespace mlir {
+namespace heir {
+
+#define GEN_PASS_DEF_WRAPGENERIC
+#include "include/Transforms/Secretize/Passes.h.inc"
+
+struct WrapWithGeneric : public OpRewritePattern<func::FuncOp> {
+  WrapWithGeneric(mlir::MLIRContext *context)
+      : mlir::OpRewritePattern<func::FuncOp>(context) {}
+
+  LogicalResult matchAndRewrite(func::FuncOp op,
+                                PatternRewriter &rewriter) const override {
+    bool hasSecrets = false;
+
+    SmallVector<Type, 4> newInputs;
+    for (unsigned i = 0; i < op.getNumArguments(); i++) {
+      if (op.getArgAttr(i, secret::SecretDialect::kArgSecretAttrName) !=
+          nullptr) {
+        hasSecrets = true;
+        op.removeArgAttr(i, secret::SecretDialect::kArgSecretAttrName);
+
+        auto newTy = secret::SecretType::get(op.getArgument(i).getType());
+        op.getArgument(i).setType(newTy);  // Updates the block argument type.
+        newInputs.push_back(newTy);
+      }
+    }
+
+    if (!hasSecrets) {
+      // Match failure, no secret inputs.
+      return failure();
+    }
+
+    llvm::SmallVector<Type, 4> newOutputs;
+    for (unsigned i = 0; i < op.getNumResults(); i++) {
+      newOutputs.push_back(secret::SecretType::get(op.getResultTypes()[i]));
+    }
+
+    op.setFunctionType(
+        FunctionType::get(getContext(), {newInputs}, {newOutputs}));
+
+    // Create a secret.generic op and pull the original function block in.
+    Block &opEntryBlock = op.getRegion().front();
+    rewriter.setInsertionPointToStart(&opEntryBlock);
+    auto newGeneric = rewriter.create<secret::GenericOp>(
+        op.getLoc(), op.getArguments(), newOutputs,
+        [&](OpBuilder &b, Location loc, ValueRange blockArguments) {
+          //  Map the input values to the block arguments.
+          IRMapping mp;
+          for (unsigned i = 0; i < blockArguments.size(); ++i) {
+            mp.map(opEntryBlock.getArgument(i), blockArguments[i]);
+          }
+          for (auto &entryOp : opEntryBlock.getOperations()) {
+            b.clone(entryOp, mp);
+          }
+          auto *returnOp = b.getBlock()->getTerminator();
+          b.create<secret::YieldOp>(loc, returnOp->getOperands());
+          returnOp->erase();
```

Comment:
I played around with `block.moveBefore`, but seemed to have trouble getting it to work

Reply:
I think your way is actually the best way because the generic hasn't been constructed yet at the time you want to do the moving. For example, I tried this, in which I tried giving access to the region pointer being constructed so that I could clone into the region. But it failed because the `cloneInto` was expecting the region input to have a terminator.

```
diff --git a/include/Dialect/Secret/IR/SecretOps.td b/include/Dialect/Secret/IR/SecretOps.td
index 6d7cced..a9fe08a 100644
--- a/include/Dialect/Secret/IR/SecretOps.td
+++ b/include/Dialect/Secret/IR/SecretOps.td
@@ -134,12 +134,18 @@ def Secret_GenericOp : Secret_Op<"generic", [
       CArg<"ValueRange">:$inputs,
       CArg<"TypeRange">:$outputTypes,
       CArg<"function_ref<void(OpBuilder &, Location, ValueRange)>">:$bodyBuilder
+    )>,
+    OpBuilder<(ins
+      CArg<"ValueRange">:$inputs,
+      CArg<"TypeRange">:$outputTypes,
+      CArg<"function_ref<void(OpBuilder &, Location, ValueRange, Region *)>">:$bodyBuilder
     )>
   ];
 
   let extraClassDeclaration = [{
     /// Defining the function type we use for building the body of affine.for.
     using BodyBuilderFn = function_ref<void(OpBuilder &, Location, ValueRange)>;
+    using BodyBuilderWithRegionFn = function_ref<void(OpBuilder &, Location, ValueRange, Region *region)>;
   }];
 
   // let hasCanonicalizer = 1;
diff --git a/lib/Dialect/Secret/IR/SecretOps.cpp b/lib/Dialect/Secret/IR/SecretOps.cpp
index f66c302..57073ce 100644
--- a/lib/Dialect/Secret/IR/SecretOps.cpp
+++ b/lib/Dialect/Secret/IR/SecretOps.cpp
@@ -217,6 +217,17 @@ void RevealOp::build(OpBuilder &builder, OperationState &result,
 void GenericOp::build(OpBuilder &builder, OperationState &result,
                       ValueRange inputs, TypeRange outputTypes,
                       GenericOp::BodyBuilderFn bodyBuilder) {
+  build(builder, result, inputs, outputTypes,
+        [bodyBuilder](OpBuilder &b, Location loc, ValueRange blockArguments,
+                      Region *region) {
+          // just ignore the region arg
+          bodyBuilder(b, loc, blockArguments);
+        });
+}
+
+void GenericOp::build(OpBuilder &builder, OperationState &result,
+                      ValueRange inputs, TypeRange outputTypes,
+                      GenericOp::BodyBuilderWithRegionFn bodyBuilder) {
   for (Type ty : outputTypes) result.addTypes(ty);
   result.addOperands(inputs);
 
@@ -231,7 +242,7 @@ void GenericOp::build(OpBuilder &builder, OperationState &result,
 
   OpBuilder::InsertionGuard guard(builder);
   builder.setInsertionPointToStart(&bodyBlock);
-  bodyBuilder(builder, result.location, bodyBlock.getArguments());
+  bodyBuilder(builder, result.location, bodyBlock.getArguments(), bodyRegion);
 }
 
 }  // namespace secret
diff --git a/lib/Transforms/Secretize/WrapGeneric.cpp b/lib/Transforms/Secretize/WrapGeneric.cpp
index 721976b..8ddeec1 100644
--- a/lib/Transforms/Secretize/WrapGeneric.cpp
+++ b/lib/Transforms/Secretize/WrapGeneric.cpp
@@ -54,15 +54,13 @@ struct WrapWithGeneric : public OpRewritePattern<func::FuncOp> {
     rewriter.setInsertionPointToStart(&opEntryBlock);
     auto newGeneric = rewriter.create<secret::GenericOp>(
         op.getLoc(), op.getArguments(), newOutputs,
-        [&](OpBuilder &b, Location loc, ValueRange blockArguments) {
+        [&](OpBuilder &b, Location loc, ValueRange blockArguments, Region *region) {
           //  Map the input values to the block arguments.
           IRMapping mp;
           for (unsigned i = 0; i < blockArguments.size(); ++i) {
             mp.map(opEntryBlock.getArgument(i), blockArguments[i]);
           }
-          for (auto &entryOp : opEntryBlock.getOperations()) {
-            b.clone(entryOp, mp);
-          }
+          op.getRegion().cloneInto(region, mp);
           auto *returnOp = b.getBlock()->getTerminator();
           b.create<secret::YieldOp>(loc, returnOp->getOperands());
           returnOp->erase();
```

Reply:
> I think your way is actually the best way because the generic hasn't been constructed yet at the time you want to do the moving. 

ah! when I tried doing it after too, `moveBefore` doesn't allow for a nice IR mapping to the new block arguments, so that was pretty annoying as well.

Thank you!

Maybe one day we'll look back at it and realize there was a one-liner. For no, it works :)